README.md
# Elmjutsu

A bag of tricks for developing with [Elm](http://elm-lang.org).

https://atom.io/packages/elmjutsu

## Setup

* Install [Elm](http://elm-lang.org/install).
* Install [language-elm](https://atom.io/packages/language-elm) (no need to install `elm-oracle` or `goto`).
* For `Error Highlighting`:
  * Install [atom-ide-ui](https://atom.io/packages/atom-ide-ui) or [linter](https://atom.io/packages/linter) + [linter-ui-default](https://atom.io/packages/linter-ui-default).
  * With linter v2 API (linter 2.3.0 and above), tooltips are collapsed by default. You have to click expand buttons (">"), or bind `linter-ui-default:expand-tooltip` to some key and hold it to expand them.
  * When moving from `atom-ide-ui` to `linter` (and vice versa), you need to reload Atom for now.
* For `Go to Definition`:
  * Install [atom-ide-ui](https://atom.io/packages/atom-ide-ui) or [hyperclick](https://atom.io/packages/hyperclick) (both optional).
* For `Autocomplete`:
  * Install [autocomplete-plus](https://atom.io/packages/autocomplete-plus) (installed by default).
  * Install [snippets](https://atom.io/packages/snippets) (optional, installed by default).
* For `Datatips`:
  * Install [atom-ide-ui](https://atom.io/packages/atom-ide-ui) (optional).
* Add your [keybindings](#keybindings).

## Features Overview

Note: Features marked with `*` are disabled by default.  You may enable them in the Settings view.

* [Error Highlighting](#error-highlighting)
  * [Always Compile Main](#always-compile-main)
  * [Set Main Paths](#set-main-paths)
  * [Set Compile Output Path](#set-compile-output-path)
  * [Report Warnings](#report-warnings)
  * [Show Inferred Type Annotations](#show-inferred-type-annotations)
  * [Compile With Debugger](#compile-with-debugger)
  * [Compile With Optimizations](#compile-with-optimizations)
  * [Quick Fixes](#quick-fixes)
  * [Quick Fix File](#quick-fix-file)

* [Autocomplete](#autocomplete)
  * [Global Autocomplete](#global-autocomplete) `*`
  * [Filtering suggestions](#autocomplete-filtering)
    * [Fuzzy Filtering](#autocomplete-fuzzy-filtering) `*`
    * [Regex Filtering](#autocomplete-regex-filtering)
    * [Type Filtering](#autocomplete-type-filtering)
  * [Autocomplete Snippets](#autocomplete-snippets) `*`
  * [Special Completions](#special-completions)
  * [Type-Aware Autocomplete](#type-aware-autocomplete) `*`
  * [Performance Tuning](#autocomplete-performance-tuning)

* Navigation
  * [Go to Definition](#go-to-definition)
  * [Go to Symbol](#go-to-symbol)
  * [Find Usages](#find-usages)
  * [Go to Next Usage](#go-to-next-usage)
  * [Go to Previous Usage](#go-to-previous-usage)
  * [Go Back](#go-back)

* Information
  * [Datatips](#datatips)
  * [Signature Help](#signature-help)
  * [Show Types in Tooltip](#show-types-in-tooltip) `*`
  * [Toggle Sidekick](#toggle-sidekick)
    * Show the type and documentation for the symbol at cursor position in a panel.
  * [Infer Type](#infer-type)
  * [Infer Types on the Fly](#infer-types-on-the-fly) `*`

* Imports Management
  * [Add Import](#add-import)
  * [Add Import As](#add-import-as)

* Package Management
  * [Install Package](#install-package)
  * [Uninstall Package](#uninstall-package)

* [Hot Reloading](#hot-reloading)

* Refactoring
  * [Rename Symbol](#rename-symbol)
  * [Surround with `let`](#surround-with-let)
  * [Lift to `let`](#lift-to-let)
  * [Lift to top-level](#lift-to-top-level)

## Feature Details

This package parses your projects' source files to extract information, and downloads documentation of 3rd-party Elm packages.  The downloaded documentation files will be saved to the path set in `Cache Directory` in the Settings view.  If `Cache Directory` is blank, a temporary directory will be used.  Take note that most operating systems delete temporary directories at bootup or at regular intervals.

### <a name="error-highlighting"></a>Error Highlighting
* This only works for Elm 0.19 onwards.  It's recommended to disable [linter-elm-make](https://atom.io/packages/linter-elm-make).
* Make sure that `Elm Path` is properly configured in the package settings.  The default works for most cases.
* You can configure the `Elm Test Path` setting to use [`elm-test`](https://github.com/rtfeldman/node-test-runner) to check for compile errors in test files (`.elm` files inside the `tests` directory).  NOTE: This only works with `elm-test@beta` (`0.19-beta10`) right now.
* If you are using Elm 0.18, it's better to disable this option by setting `Run elm make` to "never" and use [linter-elm-make](https://atom.io/packages/linter-elm-make) instead.
* If you want to change the appearance of the error messages, look into [error-highlighting.less](https://github.com/halohalospecial/atom-elmjutsu/blob/master/styles/error-highlighting.less) and copy the styles to your `~/.atom/styles.less` (where `~/.atom` is your Atom directory).

![error-highlighting](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/error-highlighting.png?raw=true)

* Useful Commands
  * Atom IDE
    * `Diagnostics: Go To Next Diagnostic`
    * `Diagnostics: Go To Previous Diagnostic`
    * `Diagnostics: Toggle Table`
  * Linter v2
    * `Linter Ui Default: Next`
    * `Linter Ui Default: Previous`
    * `Linter Ui Default: Toggle Active Editor`
  * Linter v1
    * `Linter: Next Error`
    * `Linter: Previous Error`
    * `Linter: Toggle`

#### <a name="always-compile-main"></a>Always Compile Main

* If enabled, the main file(s) will always be compiled instead of the file being edited.  The main files can be set using the `Elmjutsu: Set Main Paths` command.  If not set, the linter will look for `Main.elm` files in the source directories.  Take note that if this is enabled, modules unreachable from the main modules will not be linted.

* The main paths will be ignored if the project type is `"package"` (in `elm.json`).

* Disabled by default.

#### <a name="set-main-paths"></a>`Elmjutsu: Set Main Paths`
Sets the main paths of the project and saves them to `elmjutsu-config.json`.

Example:
```
{
  "mainPaths": ["Todo.elm", "Test.elm"]
}
```
The main paths are only relevant if `Always Compile Main` is enabled.  See [Always Compile Main](#always-compile-main).

#### <a name="set-compile-output-path"></a>`Elmjutsu: Set Compile Output Path`
Sets the output path of the project and saves it to `elmjutsu-config.json`.  Defaults to "/dev/null" if not set.

Example:
```
{
  "outputPath": "main.js"
}
```

#### <a name="report-warnings"></a>Report Warnings

* Also report warnings instead of just the errors.  Disabled by default (for now).
* Since the latest compiler (0.19) does not output warnings anymore, hacks are currently used to determine missing type annotations.

#### <a name="show-inferred-type-annotations"></a>Show Inferred Type Annotations

* This will only work if `Report Warnings` is also checked.

![show-inferred-type-annotations](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/show-inferred-type-annotations.gif?raw=true)

#### <a name="compile-with-debugger"></a>Compile With Debugger
* Adds the `--debug` flag to `elm make`.
* Enabling this will automatically disable `Compile With Optimizations`.

#### <a name="compile-with-optimizations"></a>Compile With Optimizations
* Adds the `--optimize` flag to `elm make`.
* Enabling this will automatically disable `Compile With Debugger`.

#### <a name="quick-fixes"></a>Quick Fixes

* If you have [atom-ide-ui](https://atom.io/packages/atom-ide-ui) installed, you can invoke the `Diagnostics: Show Actions At Position` command or click on the appropriate button from the panel. You can uncheck `Enable Code Actions` in the Settings view if you do not want to show the buttons.

![quick-fix-replace-with](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/quick-fix-replace-with.gif?raw=true)

* Available Quick Fixes:
  * Replace with
  * Add missing patterns
  * Add import
  * Add import as
  * Define top-level
  * Install package
  * Add type annotation
  * Convert using

#### <a name="quick-fix-file"></a>`Elmjutsu: Quick Fix File`

* Fixes all issues in the active text editor in one go. If there is more than one fix for an issue, it will choose the first from the list.

### <a name="autocomplete"></a>Autocomplete

This provides suggestions for imports, project symbols, and 3rd-party package symbols.

* It's recommended to uncheck `Enable autocomplete` of the [language-elm](https://atom.io/packages/language-elm) package to prevent duplicate suggestions.

![autocomplete](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/autocomplete.gif?raw=true)

#### <a name="global-autocomplete"></a>Global Autocomplete `*`
* Check `Enable Global Autocomplete` if you want to include unimported project symbols.
  * This will also allow :zap:`Auto import` completion (which works like [Add Import](#elmjutsu-add-import)).
  * Take note that you may experience lag if you have a large project.

![auto-import](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/auto-import.gif?raw=true)

#### <a name="autocomplete-filtering"></a>Filtering suggestions

* <a name="autocomplete-fuzzy-filtering"></a>Fuzzy Filtering `*`

  You can check `Enable Autocomplete Fuzzy Filtering` to filter suggestions using [fuzz-aldrin-plus](https://github.com/jeancroy/fuzz-aldrin-plus).

  ![autocomplete-fuzzy-filtering](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/autocomplete-fuzzy-filtering.gif?raw=true)

* <a name="autocomplete-regex-filtering"></a>Regex Filtering

  If the typed text starts with a slash (`/`), the rest of the characters will be used as a regular expression to filter the suggestions by name.

  ![autocomplete-regex-filtering](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/autocomplete-regex-filtering.gif?raw=true)

* <a name="autocomplete-type-filtering"></a>Type Filtering

  If the typed text starts with a colon (`:`), the rest of the characters will be used as a regular expression to filter the suggestions by type signature.

  ![autocomplete-type-filtering](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/autocomplete-type-filtering.gif?raw=true)

  * If you want to filter by name and type signature, put two underscores (`__`) in between.  For example, `:cons__String` will suggest `String.cons` (`Char -> String -> String`) and `String.uncons` (`String -> Maybe.Maybe ( Char, String )`).

  * Use underscores in lieu of spaces (e.g. to match `List a`, type `List_a`).

  * Remember to escape dots, vertical bars, braces (for records), parentheses (for tuples), etc.

#### <a name="autocomplete-snippets"></a>Autocomplete Snippets `*`
* You can also check `Enable Autocomplete Snippets` if you prefer.

![autocomplete-snippet](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/autocomplete-snippet.gif?raw=true)

#### <a name="special-completions"></a>Special Completions :zap:

Press <kbd>tab</kbd> to go to the next tab stop (similar to how snippets work).  Special completions can be disabled individually in the package settings.

* :zap:`Insert program`

  Type `"html"`, `"Html.program"`, `"platform"`, or `"Platform.program"` in an empty editor to insert skeleton code.

![construct-html-program](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-html-program.gif?raw=true)

* :zap:`Insert module`
* :zap:`Insert let/in`
* :zap:`Insert if/then/else`

![construct-basic](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-basic.gif?raw=true)

* :zap:`Insert case/of`

![construct-case-of](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-case-of.gif?raw=true)

* :zap:`Insert default arguments`

![construct-default-arguments](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-default-arguments.gif?raw=true)

* :zap:`Replace type with default`

![construct-default-value-for-type](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-default-value-for-type.gif?raw=true)

* :zap:`Define from type annotation`

![construct-from-type-annotation-1](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-from-type-annotation-1.gif?raw=true)

![construct-from-type-annotation-2](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-from-type-annotation-2.gif?raw=true)

![construct-from-type-annotation-3](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/construct-from-type-annotation-3.gif?raw=true)

* :zap:`Auto import`

  See [Global Autocomplete](#global-autocomplete).

* :zap:`Replace with inferred type`

  See [Type-Aware Autocomplete](#type-aware-autocomplete) for a screenshot.

#### <a name="type-aware-autocomplete"></a>Type-Aware Autocomplete `*`
* Check `Enable Type-Aware Autocomplete` if you want to prioritize suggestions matching the expected type at cursor position.
  * The type can be inferred via the `Infer Type` command, but it's recommended to check `Infer Expected Type At Cursor On The Fly` in the package settings instead.
  * WARNING: This is highly experimental and may cause lag, especially if `Enable Global Autocomplete` is also checked.

![type-aware-autocomplete](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/type-aware-autocomplete.gif?raw=true)

#### <a name="autocomplete-performance-tuning"></a>Performance Tuning

  * It's recommended to uncheck `Enable autocomplete` of the [language-elm](https://atom.io/packages/language-elm) package.  No need to install `elm-oracle`.

  * If you are experiencing lag while typing, you can set the value of `Max Number of Autocomplete Suggestions` to a small number such as `50` so that Atom will have less items to render.

  * Enabling <b>both</b> [Global Autocomplete](#global-autocomplete) and [Type-Aware Autocomplete](#type-aware-autocomplete) will usually result to lag because there will be lots of computations involved.  This may be improved in the future.

### Navigation

  * #### <a name="go-to-definition"></a>`Elmjutsu: Go To Definition`

    ![go-to-definition](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/go-to-definition.gif?raw=true)

    - If the [atom-ide-ui](https://atom.io/packages/atom-ide-ui) or [hyperclick](https://atom.io/packages/hyperclick) package is installed, you can also check `Enable Hyperclick` to jump to definition using `Ctrl` + click / `Cmd` + click (Mac).

  * #### <a name="go-to-symbol"></a>`Elmjutsu: Go To Symbol`

    ![go-to-symbol](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/go-to-symbol.gif?raw=true)

    - You can type a `:` (colon) in the text box to also filter by type.

  * #### <a name="find-usages"></a>`Elmjutsu: Find Usages`

    ![find-usages](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/find-usages.gif?raw=true)

  * #### <a name="go-to-next-usage"></a>`Elmjutsu: Go To Next Usage`

    Moves the cursor to the position of the next usage.

  * #### <a name="go-to-previous-usage"></a>`Elmjutsu: Go To Previous Usage`

    Moves the cursor to the position of the previous usage.

  * #### <a name="go-back"></a>`Elmjutsu: Go Back`

    The current cursor position is added to a navigation stack before jumping via:
      * `Go To Definition`
      * `Go To Symbol`
      * `Go To Next Usage`
      * `Go To Previous Usage`
      * Clicking on a source path link (Sidekick and Datatip).

    Invoke this command to jump back to the previous position.

  * #### <a name="hide-usages-panel"></a>`Elmjutsu: Hide Usages Panel`

    Closes the `Usages` panel (the panel is shown after invoking `Find Usages` or `Rename Symbol`).

### <a name="datatips"></a>Datatips
Provides support for [Datatips](https://github.com/facebook-atom/atom-ide-ui/blob/master/docs/datatips.md).  The [atom-ide-ui](https://atom.io/packages/atom-ide-ui) package should be installed for this to work.

![datatips](https://github.com/halohalospecial/atom-elmjutsu/blob/master/images/datatips.gif?raw=true)

### <a name="signature-help"></a>Signature Help
Provides support for [Signature Help](https://github.com/facebook-atom/atom-ide-ui/blob/master/docs/signature-help.md).  The [atom-ide-ui](https://atom.io/packages/atom-ide-ui) package should be installed for this to work.


CHANGELOG.md
## 9.11.0
* Enable Linter v1 and v2.  Thanks to @ymtszw for the PR (#158)!

## 9.10.1
* Fix #150.  Thanks to @z5h for reporting!

## 9.10.0
* Add `Enable Hot Reloading` option.

## 9.9.3
* Fix #141.  Thanks to @mrattner and @gampleman for the help!

## 9.9.2
* Fix #147.  Thanks to @aaahrens and @johnmullany for reporting and debugging!

## 9.9.1
* Fix #146.  Thanks to @aaahrens for reporting!

## 9.9.0
* Add `Compile With Debugger` option.
* Add `Compile With Optimizations` option.
* Add `Set Compile Output Path` command.
* Add `Toggle Always Compile Main` command.
* Add `Toggle Report Warnings` command.
* Add `Toggle Compile With Debugger` command.
* Add `Toggle Compile With Optimizations` command.

## 9.8.0
* Change the default error message colors.
* Add `Convert using` quick fix.
* Ignore main paths when `Always Compile Main` is enabled and the type of the project is `"package"` (in `elm.json`) (#139).  Thanks to @ymtszw for reporting!

## 9.7.0
* Add `elmjutsu.addImportAs` to provided services.

## 9.6.0
* Add `Add import as` quick fix (#134).  Thanks to @anagrius for the suggestion!

## 9.5.0
* Add `Add Import As` (#134).  Thanks to @anagrius for the suggestion!
* Update index after installing a package (Elm 0.19).
* Fix import autocompletion issue when fuzzy filtering is enabled.
* Add workaround for node-test-runner issue.  See https://github.com/halohalospecial/atom-elmjutsu/issues/131.  Thanks to @ymtszw for the PR!

## 9.4.1
* Fix `Enhanced Elm Make Messages` issues.
* Show the raw stderr when compile fails.  Thanks to @MethodGrab for the PR!

## 9.4.0
* Add `Elmjutsu: Quick Fix File` command.

## 9.3.2
* Fix #142.  Thanks to @lucamug for reporting!

## 9.3.1
* Fix issue with `Report Warnings` not working in submodules.  Thanks to @z5h for reporting!

## 9.3.0
* Add `Report Warnings` and `Show Inferred Type Annotations` options.
* Add `Add type annotation` quick fix.
* Make `Eval` work in Elm 0.19.

## 9.2.0
* Add `Define top-level` quick fix.
* Add `Install package` quick fix.

## 9.1.0
* Add `Add import` quick fix.

## 9.0.4
* Offer to run `elm init` when there's no `elm.json` or `elm-package.json` file.
* Add action buttons to compile error notifications.
* Fix `Install Package` and `Uninstall Package`.
* Make `Install Package` work with Elm 0.19.

## 9.0.3
* Fix autocompletion and `Go To Definition` for Elm 0.18 projects (#138).  Thanks to @anagrius for reporting!

## 9.0.2
* Add `node_modules` to the ignored directories (#135).  Thanks to @anagrius for the PR!
* Fix URL for documentation files.  Thanks to @anagrius for the PR!

## 9.0.1
* Fix autocompletion in test files in `"type":"package"` projects.  Thanks to @ymtszw for reporting!

## 9.0.0
* Add `Elm Test Path` setting.  Allow relative paths for `Elm Path` and `Elm Test Path`.

## 8.6.4
* Change the default error message color to white.

## 8.6.3
* Fix issue with relative paths in Elm Make messages.  Thanks to @ymtszw for reporting!

## 8.6.2
* Follow Elm Make message line length for now.

## 8.6.1
* Fix #130.  Thanks to @ababup1192 for reporting!

## 8.6.0
* Add `Always Compile Main` option and `Elmjutsu: Set Main Paths` command (#129).  Thanks to @mm-tfx for reporting and help in debugging!

## 8.5.1
* Fix issue with relative paths in Elm Make messages.

## 8.5.0
* Make autocomplete work with `package` type Elm 0.19 projects.  Thanks to @miyamoen for reporting!
* Add `Enhanced Elm Make Messages` option.  This applies additional styling to Elm Make messages and removes code snippets to save space.

## 8.4.0
* Add `Replace with` quick fix.
* Add `Add missing patterns` quick fix.

## 8.3.3
* Display elm make errors in notification popups.

## 8.3.2
* Make inferring features work with Elm 0.19.

## 8.3.1
* Add Error Highlighting to README.

## 8.3.0
* Add `Run elm make` option for Elm 0.19 (experimental).  Runs `elm make` on save by default.  If you're still using Elm 0.18, you should set this option to "never".

## 8.2.1
* Fix #121.  Thanks to @CharlonTank and @MethodGrab for reporting!

## 8.2.0
* Add some support for Elm 0.19.  `Install Package`, `Uninstall Package`, and inferring features are not yet supported.

## 8.1.4, 8.1.3
* Fix Signature Help infinite loop again.  Thanks to @SidneyNemzer for reporting and debugging!

## 8.1.2
* Fix Signature Help infinite loop.

## 8.1.1
* Do not process files when editing remotely via Atom Teletype (for now).

## 8.1.0
* Completion of record fields is now enabled by default.

## 8.0.0
* Improve typing performance.
* Greatly improve `Go to Definition` / `Hyperclick` performance (#115).  Thanks to @smucode for reporting and debugging!
* Signature Help integration is now enabled by default.
* Add basic record fields completion (disabled by default).
* Add `Log Debug Messages` option.

## 7.3.6
* Fix Signature Help infinite loop.
* Make text in Datatips follow style guide.  Thanks to @raqystyle for the PR!

## 7.3.5
* Autocomplete: Handle nested record fields.

## 7.3.4
* Fix `Define from type annotation` bug.

## 7.3.3
* Fix `Insert default arguments` bug.

## 7.3.2
* Fix bug in getting the active top-level.

## 7.3.1
* Signature Help: Fix bug with nested parens.

## 7.3.0
* Add [Signature Help](https://github.com/facebook-atom/atom-ide-ui/blob/master/docs/signature-help.md) integration.

## 7.2.6
* When scanning a buffer range, check that `match` is not null.

## 7.2.5
* Fix #88.  Thanks to @Namek for reporting!

## 7.2.4
* Check first if the file is an Elm file before getting the project directory (optimization).

## 7.2.3
* Fix crashes related to symbolic links (#107).  Thanks to @AsaAyers for the fix!
* Update dependencies.

## 7.2.2
* Update dependencies.

## 7.2.1
* Add `Insert Platform program` special completion.

## 7.2.0
* Add `Insert Html program` special completion.

## 7.1.7
* Minor performance improvements.
* Fix encoder/decoder autocompletion indentation for lists.

## 7.1.6

lib/config.js
'use babel';

import which from 'which';

export default {
  // Compilation

  runElmMake: {
    title: 'Run `elm make` on the active file',
    description:
      'Note: If you are using Elm 0.18, set this to "never" and use https://atom.io/packages/linter-elm-make instead :)',
    type: 'string',
    // enum: ['on save', 'on the fly', 'never'],
    enum: ['on save', 'never'],
    default: 'on save',
    order: 1,
  },
  alwaysCompileMain: {
    title: 'Always Compile Main',
    description:
      'Always compile the main file(s) instead of the active file.  The main files can be set using the `Elmjutsu: Set Main Paths` command.  If not set, the linter will look for `Main.elm` files in the source directories.  Modules unreachable from the main modules will not be linted.  Main paths will be ignored if the project type is `"package"` (in `elm.json`).',
    type: 'boolean',
    default: false,
    order: 2,
  },
  reportWarnings: {
    title: 'Report Warnings',
    description: 'Also report warnings instead of just the errors.',
    type: 'boolean',
    default: false, // TODO: Change default to true when compiler warnings are available again.
    order: 3,
  },
  showInferredTypeAnnotations: {
    title: 'Show Inferred Type Annotations',
    description:
      'Note: This will only work if `Report Warnings` is also checked.',
    type: 'boolean',
    default: true,
    order: 4,
  },
  compileWithDebugger: {
    title: 'Compile With Debugger',
    description:
      'Adds the `--debug` flag to `elm make`.  Enabling this will automatically disable `Compile With Optimizations`.',
    type: 'boolean',
    default: false,
    order: 5,
  },
  compileWithOptimizations: {
    title: 'Compile With Optimizations',
    description:
      'Adds the `--optimize` flag to `elm make`.  Enabling this will automatically disable `Compile With Debugger`.',
    type: 'boolean',
    default: false,
    order: 6,
  },

  // Diagnostics

  codeActionsEnabled: {
    title: 'Enable Code Actions',
    description: 'Show quick fixes as code actions (for `atom-ide-ui`).',
    type: 'boolean',
    default: true,
    order: 7,
  },
  enhancedElmMakeMessages: {
    title: 'Enhanced Elm Make Messages',
    description:
      'Apply additional styling to Elm Make messages and removes code snippets to save space.',
    type: 'boolean',
    default: true,
    order: 8,
  },

  // Autocompletion

  autocompleteEnabled: {
    title: 'Enable Autocomplete',
    type: 'boolean',
    default: true,
    order: 9,
  },
  globalAutocompleteEnabled: {
    title: 'Enable Global Autocomplete',
    description: 'Whether to include unimported project symbols.',
    type: 'boolean',
    default: false,
    order: 10,
  },
  typeAwareAutocompleteEnabled: {
    title: 'Enable Type-Aware Autocomplete',
    description:
      'WARNING: This is highly experimental and may cause lag, especially if `Enable Global Autocomplete` is also checked.',
    type: 'boolean',
    default: false,
    order: 11,
  },
  autocompleteSnippetsEnabled: {
    title: 'Enable Autocomplete Snippets',
    type: 'boolean',
    default: false,
    order: 12,
  },
  autocompleteFuzzyFilteringEnabled: {
    title: 'Enable Autocomplete Fuzzy Filtering',
    type: 'boolean',
    default: false,
    order: 13,
  },
  autocompleteMaxSuggestions: {
    title: 'Max Number of Autocomplete Suggestions',
    description:
      'No limit if set to `0` (default).  Set this to a small number (e.g. `50`) if you are experiencing lag so that Atom will have less items to render.',
    type: 'integer',
    default: 0,
    order: 14,
  },
  autocompleteDescriptionDisplay: {
    title: 'Autocomplete Description Display',
    type: 'string',
    enum: ['markdown', 'text', 'none'],
    default: 'markdown',
    order: 15,
  },

  // Special completions

  autoImportCompletionEnabled: {
    title: 'Enable `Auto import` special completion',
    type: 'boolean',
    default: true,
    order: 16,
  },
  replaceWithInferredTypeCompletionEnabled: {
    title: 'Enable `Replace with inferred type` special completion',
    type: 'boolean',
    default: true,
    order: 17,
  },
  insertProgramCompletionEnabled: {
    title: 'Enable `Insert program` special completion',
    description:
      'This will activate by typing `html`, `Html.program`, `platform`, or `Platform.program` in an empty editor.',
    type: 'boolean',
    default: true,
    order: 18,
  },
  insertModuleCompletionEnabled: {
    title: 'Enable `Insert module` special completion',
    type: 'boolean',
    default: true,
    order: 19,
  },
  insertLetInCompletionEnabled: {
    title: 'Enable `Insert let/in` special completion',
    type: 'boolean',
    default: true,
    order: 20,
  },
  insertIfThenElseCompletionEnabled: {
    title: 'Enable `Insert if/then/else` special completion',
    type: 'boolean',
    default: true,
    order: 21,
  },
  insertCaseOfCompletionEnabled: {
    title: 'Enable `Insert case/of` special completion',
    type: 'boolean',
    default: true,
    order: 22,
  },
  insertDefaultArgumentsCompletionEnabled: {
    title: 'Enable `Insert default arguments` special completion',
    type: 'boolean',
    default: true,
    order: 23,
  },
  replaceTypeWithDefaultCompletionEnabled: {
    title: 'Enable `Replace type with default` special completion',
    type: 'boolean',
    default: true,
    order: 24,
  },
  defineFromTypeAnnotationCompletionEnabled: {
    title: 'Enable `Define from type annotation` special completion',
    type: 'boolean',
    default: true,
    order: 25,
  },
  // replaceTypeWithAliasCompletionEnabled: {
  //   title: 'Enable `Replace type with alias` special completion',
  //   type: 'boolean',
  //   default: true,
  //   order: 26
  // },

  recordFieldsCompletionEnabled: {
    title: 'Enable completion of record fields',
    type: 'boolean',
    default: true,
    order: 27,
  },

  autocompleteMinChars: {
    title: 'Autocomplete Min Chars',
    description:
      'Minimum number of characters to type before showing suggestions.',
    type: 'integer',
    default: 1,
    order: 28,
  },

  // Atom IDE integration

  hyperclickEnabled: {
    title: 'Enable Hyperclick',
    type: 'boolean',
    default: true,
    order: 29,
  },
  datatipsEnabled: {
    title: 'Enable Datatips',
    type: 'boolean',
    default: true,
    order: 30,
  },
  signatureHelpEnabled: {
    title: 'Enable Signature Help',
    type: 'boolean',
    default: true,
    order: 31,
  },

  // Tooltip

  showTypesInTooltip: {
    title: 'Show Types in Tooltip',
    type: 'boolean',
    default: false,
    order: 32,
  },
  // showAliasesOfTypesInTooltip: {
  //   title: 'Show Aliases of Types in Tooltip',
  //   type: 'boolean',
  //   default: false,
  //   order: 33
  // },
  typesTooltipPosition: {
    title: 'Types Tooltip Position',
    type: 'string',
    enum: ['top', 'right', 'bottom', 'left'],
    default: 'top',
    order: 34,
  },

  // Sidekick

  showSidekick: {
    title: 'Show Sidekick',
    description: 'Show types and documentation in another panel.',
    type: 'boolean',
    default: false,
    order: 35,
  },
  showTypesInSidekick: {
    title: 'Show Types in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 36,
  },
  showTypeCasesInSidekick: {
    title: 'Show Tags of Union Types in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 37,
  },
  showDocCommentsInSidekick: {
    title: 'Show Doc Comments in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 38,
  },
  showAssociativitiesInSidekick: {
    title: 'Show Associativities in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 39,
  },
  showPrecedencesInSidekick: {
    title: 'Show Precedences in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 40,
  },
  // showAliasesOfTypesInSidekick: {
  //   title: 'Show Aliases of Types in Sidekick',
  //   type: 'boolean',
  //   default: false,
  //   order: 41
  // },
  showSourcePathsInSidekick: {
    title: 'Show Source Paths in Sidekick and Datatips',
    type: 'boolean',
    default: true,
    order: 42,
  },
  sidekickPosition: {
    title: 'Sidekick Position',
    type: 'string',
    enum: ['top', 'right', 'bottom', 'left'],
    default: 'bottom',
    order: 43,
  },
  sidekickSize: {
    title: 'Sidekick Size (in pixels)',
    description: 'Automatically resizes to fit content if set to `0`.',
    type: 'integer',
    default: 0,
    minimum: 0,
    order: 44,
  },

  // Infer

  inferExpectedTypeAtCursorOnTheFly: {
    title: 'Infer Expected Type At Cursor On The Fly',
    type: 'boolean',
    default: false,
    order: 45,
  },
  inferTypeOfSelectionOnTheFly: {
    title: 'Infer Type Of Selection On The Fly',
    type: 'boolean',
    default: false,
    order: 46,
  },
  // inferHoleTypesOnTheFly: {
  //   title: 'Infer Hole Types On The Fly',
  //   type: 'boolean',
  //   default: false,
  //   order: 47
  // },

  showScopeBar: {
    title: 'Show Scope Bar',
    type: 'boolean',
    default: false,
    order: 48,
  },

  elmExecPath: {
    title: 'Elm Path',
    description:
      'Path to the `elm` executable.  This can be an absolute path or relative to the directory of `elm.json`.',
    type: 'string',
    default: which.sync('elm', { nothrow: true }) || '',
    order: 49,
  },
  elmTestExecPath: {
    title: 'Elm Test Path',
    description:
      'Path to the `elm-test` executable.  This will used to check for errors in test files.  This can be an absolute path or relative to the directory of `elm.json`.',
    type: 'string',
    default: which.sync('elm-test', { nothrow: true }) || '',
    order: 50,
  },

  // For Elm version <= 0.18

  elmMakeExecPath: {
    title: 'Elm Make Path (for Elm version <= 0.18.0)',
    description: 'Path to the `elm-make` executable.',
    type: 'string',
    default: which.sync('elm-make', { nothrow: true }) || '',
    order: 51,
  },
  elmPackageExecPath: {
    title: 'Elm Package Path (for Elm version <= 0.18.0)',
    description: 'Path to the `elm-package` executable.',
    type: 'string',
    default: which.sync('elm-package', { nothrow: true }) || '',
    order: 52,
  },
  elmReplExecPath: {
    title: 'Elm REPL Path (for Elm version <= 0.18.0)',
    description: 'Path to the `elm-repl` executable.',
    type: 'string',
    default: which.sync('elm-repl', { nothrow: true }) || '',
    order: 53,
  },

  hotReloadingEnabled: {
    title: 'Enable Hot Reloading',
    type: 'boolean',
    default: false,
    order: 54,
  },
  hotReloadingHost: {
    title: 'Hot Reloading Host',
    type: 'string',
    default: 'localhost',
    order: 55,
  },
  hotReloadingPort: {
    title: 'Hot Reloading Port',
    type: 'integer',
    default: 0,
    order: 56,
  },

  cacheDirectory: {
    title: 'Cache Directory (for Elm version <= 0.18.0)',
    description:
      'Directory where the downloaded docs will saved.  If blank, a temporary directory will be used.  Take note that most operating systems delete temporary directories at bootup or at regular intervals.',
    type: 'string',
    default: '',
    order: 57,
  },

  evalPreludePath: {
    title: 'Eval Prelude Path',
    // description: 'Path to the prelude file for `Elmjutsu: Pipe Selections`.  This will also be used when invoking `Elmjutsu: Eval` in a non-Elm editor.  If blank, the default will be used (EvalPrelude.elm).',
    description:
      'Path to the prelude file for `Elmjutsu: Pipe Selections`.  If blank, the default will be used (EvalPrelude.elm).',
    type: 'string',
    default: '',
    order: 58,
  },

  logDebugMessages: {
    title: 'Log Debug Messages',
    description: 'Show debug messages using `console.log`.',
    type: 'boolean',
    default: false,
    order: 59,
  },
};
